akka-remote/src/main/scala/akka/remote/artery/tcp/SSLEngineProvider.scala

```diff
@@ -0,0 +1,331 @@
+    createSSLEngine(akka.stream.Server, hostname, port)
+
+  override def createClientSSLEngine(hostname: String, port: Int): SSLEngine =
+    createSSLEngine(akka.stream.Client, hostname, port)
+
+  private def createSSLEngine(role: TLSRole, hostname: String, port: Int): SSLEngine = {
+    createSSLEngine(getOrCreateContext(), role, hostname, port)
+  }
+
+  private def createSSLEngine(
+    sslContext: SSLContext,
+    role:       TLSRole,
+    hostname:   String,
+    port:       Int,
+    closing:    TLSClosing = IgnoreComplete): SSLEngine = {
+
+    val engine = sslContext.createSSLEngine(hostname, port)
+    engine.setUseClientMode(role == akka.stream.Client)
+    engine.setEnabledCipherSuites(SSLEnabledAlgorithms.toArray)
+    engine.setEnabledProtocols(Array(SSLProtocol))
+
+    if ((role != akka.stream.Client) && SSLRequireMutualAuthentication)
+      engine.setNeedClientAuth(true)
+
+    engine
+  }
+
+  override def verifyClientSession(hostname: String, session: SSLSession): Option[Throwable] =
+    verifyHostname(hostname, session)
+
+  override def verifyServerSession(hostname: String, session: SSLSession): Option[Throwable] =
+    verifyHostname(hostname, session)
+
+  private def verifyHostname(hostname: String, session: SSLSession): Option[Throwable] = {
+    if (hostnameVerifier.verify(hostname, session)) None
+    else Some(new ConnectionException(s"Hostname verification failed! Expected session to be for [$hostname]"))
+  }
+
+}
+
+/**
+ * INTERNAL API
+ *
+ * A disabled but complaining hostname verifier.
+ *
+ * Copied this from ssl-config to avoid the "dependency" and adjust logging.
+ * TODO perhaps this should be public?
+ */
+@InternalApi private[akka] class DisabledComplainingHostnameVerifier(log: MarkerLoggingAdapter) extends HostnameVerifier {
+
+  private val defaultHostnameVerifier = new DefaultHostnameVerifier(log)
+
+  override def verify(hostname: String, sslSession: SSLSession): Boolean = {
+    val hostNameMatches = defaultHostnameVerifier.verify(hostname, sslSession)
+    if (!hostNameMatches) {
+      log.warning(LogMarker.Security, "Hostname verification failed on hostname [{}], " +
+        "but the connection was accepted because hostname-verification is disabled. " +
+        "Please fix the X.509 certificate on the host to remove this warning. " +
+        "More details can be seen by enabling debug logging.", hostname)
+    }
+    true
+  }
+}
+
+/**
+ * INTERNAL API
+ *
+ * Use the internal sun hostname checker as the hostname verifier. Thanks to Kevin Locke.
+ *
+ * Copied this from ssl-config to avoid the "dependency" and adjust logging.
+ * TODO perhaps this should be public?
+ *
+ * @see sun.security.util.HostnameChecker
+ * @see http://kevinlocke.name/bits/2012/10/03/ssl-certificate-verification-in-dispatch-and-asynchttpclient/
+ */
+@InternalApi private[akka] class DefaultHostnameVerifier(log: MarkerLoggingAdapter) extends HostnameVerifier {
```

Comment:
For hostname checking as needed for HTTPS no extra code is needed any more, if [SSLParameters.setEndpointIdentificationAlgorithm("https")](https://docs.oracle.com/javase/8/docs/api/javax/net/ssl/SSLParameters.html#setEndpointIdentificationAlgorithm-java.lang.String-) is used. Using this code will mean that we depend on Oracle internal features.

So, I guess we could also just use `setEndpointIdentificationAlgorithm("https")` which should work out of the box if the SSLEngine is created with host and port. `setEndpointIdentificationAlgorithm` was only added in Java 7 so that might be a reason that ssl-config was necessary before.

It seems this code adds 1) some Kerberos support 2) better error messages when checking hostnames certificates fail.

It seem this code original comes from a [blog post](https://tersesystems.com/blog/2014/03/23/fixing-hostname-verification/) of @wsargent. It seems to imply that the solution was necessary for AsyncHttpClient for which the Java 7 solution didn't work.

I think one way going forward would be this:

 * remove this checker
 * add a config setting that if enabled would call `setEndpointIdentificationAlgorithm("https")` and has an empty implementation of the verify methods
 * if you need more support than this you have to implement a custom `SSLEngineProvider`
 * we might want to add a ticket to improve error messages if hostname checking fails in our TLS stage

WDYT?

Reply:
I think it would be good to use that built in way instead of the custom code. It seems to work with:

```
    val engine = sslContext.createSSLEngine(hostname, port)
    val sslParams = sslContext.getDefaultSSLParameters
    sslParams.setEndpointIdentificationAlgorithm("HTTPS")
    engine.setSSLParameters(sslParams)
```

Is that the right way?

I guess we have to have hostname verification disabled by default. Should we anyway log a warning message recommending to turn it on?

btw, we better use upper case HTTPS as that is what is specified in https://docs.oracle.com/javase/7/docs/technotes/guides/security/StandardNames.html#jssenames

Reply:
Added this ticket https://github.com/akka/akka/issues/24571 for better error logging in TLS.